db/functions/base.py

```diff
@@ -0,0 +1,261 @@
+    hints = tuple([
+        hints.returns(hints.boolean),
+        hints.parameter_count(2),
+        hints.all_parameters(hints.comparable),
+    ])
+
+    @staticmethod
+    def to_sa_expression(p1, p2):
+        return p1 < p2
+
+
+class In(DbFunction):
+    id = 'in'
+    name = 'In'
+    hints = tuple([
+        hints.returns(hints.boolean),
+        hints.parameter_count(2),
+        hints.parameter(2, hints.array),
+    ])
+
+    @staticmethod
+    def to_sa_expression(p1, p2):
+        return p1.in_(p2)
+
+
+class And(DbFunction):
+    id = 'and'
+    name = 'And'
+    hints = tuple([
+        hints.returns(hints.boolean),
+    ])
+
+    @staticmethod
+    def to_sa_expression(*ps):
+        return and_(*ps)
+
+
+class Or(DbFunction):
+    id = 'or'
+    name = 'Or'
+    hints = tuple([
+        hints.returns(hints.boolean),
+    ])
+
+    @staticmethod
+    def to_sa_expression(*ps):
+        return or_(*ps)
+
+
+class StartsWith(DbFunction):
+    id = 'starts_with'
+    name = 'Starts With'
+    hints = tuple([
+        hints.returns(hints.boolean),
+        hints.parameter_count(2),
+        hints.all_parameters(hints.string_like),
+    ])
+
+    @staticmethod
+    def to_sa_expression(p1, p2):
+        return p1.like(f'{p2}%')
+
+
+class ToLowercase(DbFunction):
+    id = 'to_lowercase'
+    name = 'To Lowercase'
+    hints = tuple([
+        hints.parameter_count(1),
+        hints.all_parameters(hints.string_like),
+    ])
+
+    @staticmethod
+    def to_sa_expression(p1):
+        return func.lower(p1)
+
+
+class ExtractURIAuthority(DbFunction):
+    id = 'extract_uri_authority'
+    name = 'Extract URI Authority'
+    hints = tuple([
+        hints.parameter_count(1),
+        hints.parameter(1, hints.uri),
+    ])
+
+    @staticmethod
+    def to_sa_expression(p1):
+        return func.getattr(URIFunction.AUTHORITY)(p1)
+
+
+def _get_defining_module_members_that_satisfy(predicate):
+    # NOTE: the value returned by globals() (when it's called within a function) is set when the
+    # function is defined and does not change depending on where the function is called from.
+    # See https://docs.python.org/3/library/functions.html#globals
+    # If we wanted to move this function into another namespace, we would have to additionally
+    # pass it this namespace's globals().
+    defining_module_name = globals()['__name__']
+    defining_module = importlib.import_module(defining_module_name)
+    all_members_in_defining_module = inspect.getmembers(defining_module)
+    return tuple(
+        member
+        for _, member in all_members_in_defining_module
+        if predicate(member)
+    )
+
+
+def _is_concrete_db_function_subclass(member):
+    return inspect.isclass(member) and member != DbFunction and issubclass(member, DbFunction)
+
+
+# Enumeration of supported DbFunction subclasses; needed when parsing.
+supported_db_functions = _get_defining_module_members_that_satisfy(_is_concrete_db_function_subclass)
```

Comment:
I like this. If generating a dict or list of `DBFunction` subclasses, that would mean they're not top-level type objects and thus can't be imported like `from db.functions.base import ToLowercase`. Might not be a problem.

This makes me wonder how I'll get `DBFunction`s defined in different namespaces collected into a single list. The functions endpoint has to be parametrized on the database because not all types and not all database functions might be installed on all databases.

Reply:
You could generate them in a dict, then (in the same loop) add them to `globals` (I've used this in other projects).  I suggest _not_ doing that, though, since most of the places you want them, you need to look them up anyway, so you may as well grab them from a dict or list or whatever.  You could import the various dicts into the `base` module (this also nicely avoids circular imports) .  You should be able to restrict which dictionaries get passed through the `DBFunction` factory based on what's installed, I think.

Reply:
Thanks for the tips. Useful.